Full batch broadcasting support for CTF 2D and 1D

reddismorr opened the pull request:
Hi, I tested your CTF creation and it looked like it is planned to be batch broadcastable. I tried it on a single set of arguments and it worked pretty well, but when I moved to a batch, it crashed. So, I have changed it a bit to allow it to work on batches of whatever shape. I also have deleted the part with additional check and dimensionality extension for spherical aberrations, because now it is assumed to be a batch variable.
I guess it would be better to add some initial tests on the shapes of the input tensors, but I also think that it might be managed outside of these functions (as repetition of the same voltage if one doesn't need to consider different values (which is never the case I guess)).
I also moved some things around in the 1D CTF to make it more readable without significant performance drops.
The overall motivation for making it batch-friendly was to use it with torch-fourier-slice and new torch-2dtm as filters. Consider different B-factors for different tilts during 2DTM with [tilts, patches, ...] form of batches.

P.S. There is a Nico Grigorieff exposure filter in dose weighting. In Relion it is applied as a substitute for a B-factor dumping, so it might be beneficial to introduce it to these CTF filters as well (I see that one can switch off B-factor dumping by setting B-factors to zero, but it might be more user-friendly to suggest dose and a key as input parameters and then select a dumping filter inside)

reddismorr commented:
Also added import for the submodules in __init__.py, just a quality-of-life commit.
BTW: Since it would be too complicated to change the code structure, I suggest renaming all instances of torch, einops to _torch and _einops in the main scrips, so they don't pop up with autocomplete.

alisterburt reviewed (commented):
Very nice! Thanks for taking the time to make a PR :-)

This looks good from my perspective, would you mind updating the typing in the docstrings to `float | torch.Tensor` rather than `float or torch.Tensor`

alisterburt commented:
looking at the tests it seems output shapes no longer match what the tests were expecting, could you also either
- update the tests to match new expectations
- update the code to match the tests if expectations were already good and the change is unexpected

alisterburt commented:
> I suggest renaming all instances of torch, einops to _torch and _einops in the main scrips, so they don't pop up with autocomplete

What do you mean by this? What you're describing is not a problem I recognise/understand

jdickerson95 commented:
Hi, thanks for the PR! It was previously only working on batches of defocus/astig and Cs before meshing them together. It was a bit clunky. I think your batched implementation is cleaner and I agree the shapes can be managed outside of these functions. 

Tests just need to be updated so it’s passing (plus Alister's docstring comment) and then I’m happy with it.

If you're interested in using these teamtomo packages, we have been working on a package for 2DTM that is now functional to the same point as cisTEMs version. It’s here: 
https://github.com/Lucaslab-Berkeley/Leopard-EM 
It has an input structure setup using pedantic classes and we’ve tried to make it as modular as possible to make it easy to add extra functionality without having to change much. We're hoping it's a useful base for people to quickly try the sorts of things you're talking about and we're open to any PRs/feedback to make it more developer-friendly. 

I'd like to keep the B-factor in the ctf for now and keep the dose weighting separate so it’s more explicit. We’re applying dose weighting to the 3D volume currently. We’re using the B-factor in the CTF instead of in ttsim3d so we don’t have to simulate a new volume to optimize the B-factor for template matching.

reddismorr commented:
Hi, Alister and  Josh,
Sorry for not checking the tests from the beginning; now, they should pass without problems. I also added a copy of the 1D CTF test for 2D CTF and checked the shapes for calls with batched input. Docstrings are also updated as you asked.
About autocomplete. Maybe I wasn't clear on that. I meant that after importing the script into Jupyter Notebook, one would also import all its imports (such as torch, einops, PackageNotFoundError, etc.) and it's fine. However, all these inner imports will pop up if you scroll through the methods and functions on this script, which sometimes might be overwhelming. Not for usage, but for development. So, as far as I know, there are two ways to prevent it:
1) Distribute large semantically connected pieces of code in separate folders with their `__init__.py`, where you select what must be imported and enlisted in the package (e.g. PyTorch, where torch.fft is a separate folder). There is also `__all__` to do a proper `from X import *`, but I am not sure how useful it is.
2) Hide all user-irrelevant imports as you hide user-irrelevant functions by adding an underscore to their name (people usually do something like `import numpy as _np`).
I did an example in ctf.py in a separate commit, so you could dismiss it if you think it is unnecessary (which is totally fine, it makes code a little bit ugly). I think it might be helpful for people like me, who usually forget what is inside the package and scroll through the autocomplete menu.
![image](https://github.com/user-attachments/assets/6cb84c80-2ec7-465e-b69d-e1022150a270)

mgiammar commented:
Thanks again for working on this PR. Actually supporting batched inputs well is a huge qualify-of-life improvement.

For adding underscores for the import naming scheme (e.g. `import numpy as _np`) I'm not aware of any other scientific programming packages taking this approach. Reorganizing the code layout to include proper `__init__.py` files with `__all__` would be my vote if you think adjusting how the autocomplete is working, BUT this would also be a large effort over all the TeamTomo packages for no improvement in functionality. However, code usability and readability comes first IMO, and I think the underscores do decrease readability.

Otherwise I think the PR looks good.

reddismorr commented:
Well, yes, I agree that the reorganization of code is going to be very time-consuming without actual functionality improvement. 
So, just to make it clear, do I need to retract my last two commits with underscores myself and make a new PR or you will do it yourself with a cherry-pick?

mgiammar commented:
If you add on one more commit which undoes the underscores in imports I think that should suffice :) 
Keeps git history linear and no shenanigans with opening other PRs.

jdickerson95 commented:
Thanks for this it looks great! @alisterburt are you also happy with everything?

alisterburt commented on src/torch_fourier_filter/__init__.py:
```diff
@@ -1,10 +1,33 @@
+    'phase_randomize',
+    'utils',
+    'whitening',
+]
```
```suggestion
]

```

alisterburt commented on src/torch_fourier_filter/ctf.py:
```diff
@@ -272,17 +276,14 @@ def calculate_ctf_1d(
+        idx = (k4 < 0).squeeze(dim=(-2,-1)) 
+        ctf[idx] *= torch.exp(k4[idx] * n4[idx])
+    ctf = einops.reduce(ctf, "... os k -> ... k", reduction="mean") # oversampling reduction
+    return ctf
```
```suggestion
    return ctf

```

alisterburt commented on tests/test_ctf.py:
```diff
@@ -89,3 +90,44 @@ def test_calculate_relativistic_electron_wavelength():
+          [ 0.3351,  0.5508,  0.9657,  0.5861, -0.8741,  0.0932, -0.8741,  0.5861,  0.9657,  0.5508]]]
+    )
+    assert (result.shape == (1,1,2,10,10))
+    assert torch.allclose(result[0,0], expected, atol=1e-4)
```
```suggestion
    assert torch.allclose(result[0,0], expected, atol=1e-4)

```

alisterburt commented:
yep all looks good to me, thanks again for taking the time here ! 

@reddismorr if you're using these packages and want to keep up on what's going on we discuss things on the image.sc zuluip [#TeamTomo](https://imagesc.zulipchat.com/#narrow/channel/426493-TeamTomo) channel and do a call quarterly 🙂
